File: lab4/src/warehouse/resources/Resource.java
package warehouse.resources;

import enums.MeasureUnit;
import request.SupplyRequestDetail;
import warehouse.WarehouseEntry;

public abstract class Resource extends WarehouseEntry<Float> {

    Float quantity;
    MeasureUnit measureUnit;

    public Resource(int id, String name, MeasureUnit measureUnit, float quantity) {
        super(id, name);
        if(quantity < 1.0f){
            throw new IllegalArgumentException("Недопустимое количество");
        }            
        this.measureUnit = measureUnit;
        this.quantity = quantity;
    }

    public MeasureUnit getMeasureUnit(){
        return measureUnit;
    }

    public float getQuantity(){
        return quantity;
    }

    @Override
    public void increase(Number value){
        quantity+=value.floatValue();
    }

    @Override
    public Float decrease(Number value){
        Float taken = Math.min(quantity, value.floatValue());
        quantity-=taken;
        return taken;
    }
    
    @Override
    public Boolean isEmpty(){
        return quantity < 1.0f;
    }

    @Override
    public abstract Boolean isMatches(SupplyRequestDetail detail);
}

File: lab5/C++/src/Resource.cpp
#include "stdexcept"
#include "Resource.hpp"
#include "WarehouseEntry.hpp"

Resource::Resource(int id, const std::string& name, MeasureUnit measureUnit, float quantity)
    : WarehouseEntry<float>(id, name), m_measureUnit(measureUnit){
    if (quantity < Resource::MIN_QUANTITY)
        throw std::invalid_argument("Недопустимое количество");
    m_quantity = quantity;
}

MeasureUnit Resource::GetMeasureUnit() const {
    return m_measureUnit;
}

float Resource::GetQuantity() const {
    return m_quantity;
}

bool Resource::IsEmpty() const {
    return m_quantity <= 0;
}

void Resource::Increase(float quantity) {
    m_quantity += quantity;
}

float Resource::Decrease(float quantity) {
    float taken = std::min(m_quantity, quantity);
    m_quantity -= taken;
    return taken;
}